server/start.go

```diff
@@ -167,9 +169,21 @@ is performed. Note, when enabled, gRPC will also be automatically enabled.
 				serverCtx.Logger.Info("starting ABCI without CometBFT")
 			}
 
-			return wrapCPUProfile(serverCtx, func() error {
+			err = wrapCPUProfile(serverCtx, func() error {
 				return start(serverCtx, clientCtx, appCreator, withCMT, opts)
 			})
+
+			serverCtx.Logger.Debug("received quit signal")
+
+			// wait for app.Close() and tmNode.Stop() to flush async db writes to disk
+			graceDuration, _ := cmd.Flags().GetDuration(FlagShutdownGrace)
+			if graceDuration > 0 {
+				serverCtx.Logger.Info("graceful shutdown start", FlagShutdownGrace, graceDuration)
+				<-time.After(graceDuration)
```

Comment:
Why do we need this, I think the defer functions will run when received signals

Reply:
Yes, defers run in all simple sample programs, however in practice, cosmos-sdk baseapp exits before defers can finish.  

Here is an example of junod with pebbledb and goleveldb with and without shutdown-grace
```
# junod exits without any database closes
# shutdown-grace = default 0

## pebbledb
^CWaitForQuitSignals() interrupt
do not wait for database to flush module=server
vvvvvvvvvvvvvvvvv baseApp defer exit module=server
--exit without db close--

## goleveldb
^CWaitForQuitSignals() interrupt
vvvvvvvvvvvvvvvvv baseApp defer exit module=server
WaitForQuitSignals() interrupt
do not wait for database to flush module=server
--exit without db close--




# junod closes db before exit with shutdown-grace
# shutdown-grace = "4.5s"

## pebbledb

commit synced commit=436F6D6D697449447B5B31373520383020323337203131312033203820343820313838203437203232372031313820323434203130372031333720353820373920333320313137203838203635203138203233382031333920323532203220373920343420343220333020313131203231362038365D3A3242423236327D module=server
committed state app_hash=AF50ED6F030830BC2FE376F46B893A4F2175584112EE8BFC024F2C2A1E6FD856 height=2863714 module=state num_txs=0
indexed block exents height=2863714 module=txindex
^CWaitForQuitSignals() interrupt
wait for database to flush graceDuration=4500 module=server
WaitForQuitSignals() interrupt
vvvvvvvvvvvvvvvvv baseApp defer exit module=server
service stop impl={"Logger":{"Logger":{}}} module=server msg={}
Stopping Node module=server
Closing blockstore module=server
Closing statestore module=server
Closing evidencestore module=server
Closing application.db module=server
Closing snapshots/metadata.db module=server
exiting... module=server
finished waiting for database to flush module=server


## goleveldb
commit synced commit=436F6D6D697449447B5B31313820323531203439203139312031323120323131203136312031303220323131203234392034342031363520313435203130302031203436203137342038203820313634203233322032323520383320313739203131302031363620392038203130382031332034342036335D3A3242423331397D module=server
committed state app_hash=76FB31BF79D3A166D3F92CA59164012EAE0808A4E8E153B36EA609086C0D2C3F height=2863897 module=state num_txs=0
indexed block exents height=2863897 module=txindex
vvvvvvvvvvvvvvvvv baseApp defer exit module=server
WaitForQuitSignals() interrupt
Closing blockstore module=server
Closing statestore module=server
Closing evidencestore module=server
Closing application.db module=server
Closing snapshots/metadata.db module=server
exiting... module=server
finished waiting for database to flush module=server
```

Reply:
I suggest we diagnose the root cause first, I've diagnosed some similar issues before, I'll have another look at what happens with our chain binary.

Reply:
Would you prefer to merge this PR without the grace period and only close the open databases?

Reply:
Hi @yihuang, any news for the root cause diagnosis?